File: p804e/unique_morse.go
package main

import "fmt"

var MORSE = []string{".-", "-...", "-.-.", "-..", ".", "..-.", "--.", "....", "..", ".---", "-.-", ".-..", "--", "-.", "---", ".--.", "--.-", ".-.", "...", "-", "..-", "...-", ".--", "-..-", "-.--", "--.."}

func uniqueMorseRepresentations(words []string) int {
	set := make(map[string]int)
	for _, word := range words {
		encoded := ""
		for _, c := range word {
			encoded += MORSE[c-'a']
		}
		set[encoded]++
	}
	return len(set)
}

type Test struct {
	words    []string
	expected int
}

func main() {
	tests := []Test{
		{
			[]string{"gin", "zen", "gig", "msg"},
			2,
		},
	}
	for _, t := range tests {
		actual := uniqueMorseRepresentations(t.words)
		fmt.Println("# of unique Morse in ", t.words, "->", actual)
	}
}

File: p804e/unique_morse.py
from typing import List


class Solution:
    MORSE = [
        ".-",
        "-...",
        "-.-.",
        "-..",
        ".",
        "..-.",
        "--.",
        "....",
        "..",
        ".---",
        "-.-",
        ".-..",
        "--",
        "-.",
        "---",
        ".--.",
        "--.-",
        ".-.",
        "...",
        "-",
        "..-",
        "...-",
        ".--",
        "-..-",
        "-.--",
        "--..",
    ]

    def uniqueMorseRepresentations(self, words: List[str]) -> int:
        morse_codes = set()
        for word in words:
            tf = [Solution.MORSE[ord(c) - 97] for c in word]
            morse_codes.add("".join(tf))
        return len(morse_codes)


# TESTS
for words, expected in [
    (["aaa"], 1),
    ([], 0),
    (["gin", "zen"], 1),
    (["gin", "zen", "gig", "msg"], 2),
]:
    sol = Solution()
    actual = sol.uniqueMorseRepresentations(words)
    print("# of different transformations in", words, "->", actual)
    assert actual == expected
